client/ayon_core/hosts/max/plugins/publish/validate_instance_in_context.py

```diff
@@ -0,0 +1,106 @@
+from pymxs import runtime as rt
+
+
+class ValidateInstanceInContext(pyblish.api.InstancePlugin,
+                                OptionalPyblishPluginMixin):
+    """Validator to check if instance asset match context asset.
+
+    When working in per-shot style you always publish data in context of
+    current asset (shot). This validator checks if this is so. It is optional
+    so it can be disabled when needed.
+
+    Action on this validator will select invalid instances.
+    """
+
+    order = ValidateContentsOrder
+    label = "Instance in same Context"
+    optional = True
+    hosts = ["max"]
+    actions = [SelectInvalidAction, RepairAction]
+
+    def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+
+        instance_node = rt.getNodeByName(instance.data.get(
+            "instance_node", ""))
+        if not instance_node:
+            return
+        asset = rt.getUserProp(instance_node, "folderPath")
+        context_asset = instance.context.data["folderPath"]
+        task = rt.getUserProp(instance_node, "task")
+        context_task = instance.context.data["task"]
+        if asset != context_asset:
+            raise PublishValidationError(
+                message=(
+                    "Instance '{}' publishes to different asset than current "
+                    "context: {}. Current context: {}".format(
+                        instance.name, asset, context_asset
+                    )
+                ),
+                description=(
+                    "## Publishing to a different asset\n"
+                    "There are publish instances present which are publishing "
+                    "into a different asset than your current context.\n\n"
+                    "Usually this is not what you want but there can be cases "
+                    "where you might want to publish into another asset or "
+                    "shot. If that's the case you can disable the validation "
+                    "on the instance to ignore it."
+                )
+            )
+        if task != context_task:
+            raise PublishValidationError(
+                message=(
+                    "Instance '{}' publishes to different task than current "
+                    "context: {}. Current context: {}".format(
+                        instance.name, task, context_task
+                    )
+                ),
+                description=(
+                    "## Publishing to a different asset\n"
+                    "There are publish instances present which are publishing "
+                    "into a different asset than your current context.\n\n"
+                    "Usually this is not what you want but there can be cases "
+                    "where you might want to publish into another asset or "
+                    "shot. If that's the case you can disable the validation "
+                    "on the instance to ignore it."
+                )
+            )
```

Comment:
I would always try and report all the issues - instead of only asset mismatch if both asset + task were wrong.
How about this?
```suggestion
        invalid = []
        if asset != context_asset:
            invalid.append(
                "Instance '{}' publishes to different asset than current "
                "context: {}. Current context: {}".format(
                    instance.name, asset, context_asset
                )
            )
        if task != context_task:
            invalid.append(
                "Instance '{}' publishes to different task than current "
                "context: {}. Current context: {}".format(
                    instance.name, task, context_task
                )
            )
        
        if invalid:
            message = "\n".join(invalid)
            raise PublishValidationError(
                message=message,
                description=(
                    "## Publishing to a different asset\n"
                    "There are publish instances present which are publishing "
                    "into a different asset or task than your current context.\n\n"
                    "Usually this is not what you want but there can be cases "
                    "where you might want to publish into another asset or "
                    "shot. If that's the case you can disable the validation "
                    "on the instance to ignore it."
                )
            )
```

Or:
```python
        invalid = False
        if asset != context_asset:
            self.log.error(
                "Instance '{}' publishes to different asset than current "
                "context: {}. Current context: {}".format(
                    instance.name, asset, context_asset
                )
            )
            invalid = True
        if task != context_task:
            self.log.error(
                "Instance '{}' publishes to different task than current "
                "context: {}. Current context: {}".format(
                    instance.name, task, context_task
                )
            )
            invalid = True
        
        if invalid:
            raise PublishValidationError(
                message="Instance '{}' does not publish into the current context".format(instance.name),
                description=(
                    "## Publishing to a different asset\n"
                    "There are publish instances present which are publishing "
                    "into a different asset or task than your current context.\n\n"
                    "Usually this is not what you want but there can be cases "
                    "where you might want to publish into another asset or "
                    "shot. If that's the case you can disable the validation "
                    "on the instance to ignore it."
                )
            )
```